cilium/CFP-23380-ordered-policy.md

```diff
@@ -0,0 +1,329 @@
+
+### Processing incremental updates
+
+In incremental update is when a complete policy newly selects, or no longer selects, one or more identities. These incremental updates must then be propagated down to the MapState and the BPF map updated accordingly.
+
+Incremental updates are aggregated by selectors and then inserted in to the MapState. The MapState is responsible for deconflicting overlapping selectors. This remains the same in the ordered version.
+
+The important change is that incremental updates must be inserted in to the map state *in priority order*. That is to say, higher-priority PerSelectorPolicies must apply their updates first. Then, the usual MapState deconfliction rules apply, and there is no risk of an incremental update diverging from policy intent.
+
+It is safe to insert incremental updates in this manner because identities and selectors are immutable. As a direct consequence, two selectors S1 and S2 might overlap, a new identity I1 can never change state from non-overlapping to overlapping.
+
+We will need to ensure that SelectorCache updates are processed in a transactional manner -- that is to say, an incremental update must be delivered to all affected selectors *before* distributing this to the MapState.
+
+#### Example 1: no wildcards
+
+Consider a set of rules:
+
+1. tcp 0-1023, ID (5,6)
+2. tcp 80, ID (7)
+3. tcp *, ID (8)
+
+Then an incremental update transaction adds ID 9. The sequence is thus:
+
+1. Rule 1 checks ID 9. ID 9 is not present in any ancestor rules. ID 9 is added.
+2. Rule 2 checks ID 9. ID 9 is present in an ancestor rule. ID 9 is **not added**.
+3. Rule 3 checks ID 9. There are no ancestor rules. ID 9 is added.
+
+
+```mermaid
+flowchart LR
+    subgraph Before
+    direction TB
+    rule1["prio 1, tcp, 0-1023, id {5, 6}"]
+    rule2["prio 2, tcp, 80, id {7}"]
+    rule3["prio 3, tcp, id {8}"]
+
+    root[[root]]
+    
+    root --> rule3
+    rule3 --> rule1
+    rule1 --> rule2
+    end
+    
+    subgraph After
+    direction TB
+    rule1a["prio 1, tcp, 0-1023, id {5, 6, 9}"]
+    rule2a["prio 2, tcp, 80, id {7, 9}"]
+    rule3a["prio 3, tcp, id {8}"]
+
+    roota[[root]]
+    
+    roota --> rule3a
+    rule3a --> rule1a
+    rule1a --> rule2a
+    end
+    
+    Before ~~~ After
+    
+    
+```
+
+#### Example 2: Wildcards
+
+
+## Impacts / Key Questions
+
+### KQ1: How should unordered and ordered policies interfact?
+
+To preserve existing behavior, all policies without an explicit order will have order number 0. Since policies with the same numerical order are merged, this preserves existing behavior.
+
+Only CCNPs are allowed to have a negative order number. This prevents namespace administrators from overriding cluster-wide policies.
+
+
```

Comment:
Might want to document the impact on `enableDefaultDeny: false`. I don't think it's an issue - the same principle that "all policies applying to the endpoint must set this to false to avoid generating the default deny", so priority doesn't interact at all. But the question could come up about how these two interact so it'd be nice to clearly define how it's not an issue.